File: experiments/plaw_raes/plot_plaw_raes/plot_trajectories_by_p.jl
using Pkg
Pkg.activate(".")

using Plots
using DelimitedFiles
using Printf


file_traj_p = "results/plaw_raes/spectral_gap_trajectory_by_p.csv"
file_baseline = "results/plaw_raes/spectral_gap_baseline_p1.csv"
output_plot = "results/plaw_raes/plot_spectral_gap_by_p.png"


if !isfile(file_traj_p)
    error("File traiettorie non trovato: $file_traj_p")
end
data_p, header_p = readdlm(file_traj_p, ',', header=true)


if !isfile(file_baseline)
    println("ATTENZIONE: File baseline non trovato ($file_baseline). Il grafico non avrà la linea di riferimento.")
    has_baseline = false
else
    data_base, header_base = readdlm(file_baseline, ',', header=true)
    has_baseline = true
end


p = plot(
    title = "Evoluzione Spectral Gap al variare di P (K=2.0)",
    xlabel = "Round",
    ylabel = "Spectral Gap Medio",
    legend = :bottomright,
    grid = :on,
    size = (900, 600),
    dpi = 300,
    ylims = (0.2, 0.65) 
)


rounds_x = 1:(size(data_p, 2) - 1)


colors = [:firebrick, :orange, :green, :cyan, :blue] 


for i in axes(data_p, 1)
    p_val = data_p[i, 1]
    gaps = data_p[i, 2:end] 
    
    col = get(colors, i, :black)
    
    plot!(p, rounds_x, gaps, 
          label = "p = $(p_val)", 
          linewidth = 2, 
          alpha = 0.8,
          color = col)
end


if has_baseline
    base_rounds = data_base[:, 1]
    base_gaps = data_base[:, 2]
    
    plot!(p, base_rounds, base_gaps, 
          label = "Baseline (p=1.0)", 
          linestyle = :dash, 
          color = :black, 
          linewidth = 2.5)
end


mkpath(dirname(output_plot))
savefig(p, output_plot)
println("Grafico salvato con successo in: $output_plot")


display(p)

File: experiments/plaw_raes/plot_plaw_raes_py/plot_trajectories_by_p.py
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


results_folder = "results/plaw_raes"
file_traj_p = os.path.join(results_folder, "spectral_gap_trajectory_by_p.csv")
file_baseline = os.path.join(results_folder, "spectral_gap_baseline_p1.csv")
output_plot = os.path.join(results_folder, "plot_trajectories_by_p_python.png")

if not os.path.exists(file_traj_p):
    raise FileNotFoundError(f"File traiettorie non trovato: {file_traj_p}")


df_p = pd.read_csv(file_traj_p)


has_baseline = False
df_base = pd.DataFrame()

if os.path.exists(file_baseline):
    df_base = pd.read_csv(file_baseline)
    has_baseline = True
else:
    print(f"ATTENZIONE: File baseline non trovato ({file_baseline}). Niente linea nera.")


sns.set_theme(style="whitegrid") 
plt.figure(figsize=(10, 6.6))    


unique_p = len(df_p)
palette = sns.color_palette("tab10", unique_p)


for i, (index, row) in enumerate(df_p.iterrows()):
    p_val = row.iloc[0]
    gaps = row.iloc[1:].values.astype(float) 
    rounds_x = range(1, len(gaps) + 1)       
    
    plt.plot(rounds_x, gaps,
             label=f"p = {p_val}",
             linewidth=2,
             alpha=0.8,
             color=palette[i % len(palette)]) 


if has_baseline:
    base_rounds = df_base.iloc[:, 0]
    base_gaps = df_base.iloc[:, 1]
    
    plt.plot(base_rounds, base_gaps,
             label="Baseline (p=1.0)",
             linestyle="--",
             color="black",
             linewidth=2.5)


plt.title("Spectral Gap Evolution by p (k=2.0)", fontsize=14)
plt.xlabel("Round", fontsize=12)
plt.ylabel("Average Spectral Gap", fontsize=12)

plt.ylim(0.2, 0.65)

plt.legend(loc="lower right", fontsize=10)
plt.tight_layout()


os.makedirs(os.path.dirname(output_plot), exist_ok=True)
plt.savefig(output_plot, dpi=300)
print(f"Grafico salvato con successo in: {output_plot}")
